libbpf-tools/memleak.c

```diff
@@ -0,0 +1,972 @@
+		goto cleanup;
+	}
+
+	// if running a specific userspace program,
+	// notify the child process that it can exec its program
+	if (strlen(env.command)) {
+		ret = event_notify(child_exec_event_fd, 1);
+		if (ret) {
+			fprintf(stderr, "failed to notify child to perform exec\n");
+
+			goto cleanup;
+		}
+	}
+
+#ifdef USE_BLAZESYM
+	symbolizer = blazesym_new();
+#endif
+
+	printf("Tracing outstanding memory allocs...  Hit Ctrl-C to end\n");
+
+	// main loop
+	while (!exiting && env.nr_intervals) {
+		env.nr_intervals--;
+
+		sleep(env.interval);
+
+		if (env.combined_only)
+			print_outstanding_combined_allocs(combined_allocs_fd, stack_traces_fd);
+		else
+			print_outstanding_allocs(allocs_fd, stack_traces_fd);
+	}
+
+	// after loop ends, check for child process and cleanup accordingly
+	if (env.pid > 0) {
+		if (!child_exited) {
+			if (kill(env.pid, SIGTERM)) {
+				perror("failed to signal child process");
+				ret = -errno;
+
+				goto cleanup;
+			}
+			printf("signaled child process\n");
+		}
+
+		if (waitpid(env.pid, NULL, 0) < 0) {
+			perror("failed to reap child process");
+			ret = -errno;
+
+			goto cleanup;
+		}
+		printf("reaped child process\n");
+	}
+
+cleanup:
+#ifdef USE_BLAZESYM
+	blazesym_free(symbolizer);
+#endif
+	memleak_bpf__destroy(skel);
+
+	free(allocs);
+	free(stack);
+
+	printf("done\n");
+
+	return ret;
+}
+
+long argp_parse_long(int key, const char *arg, struct argp_state *state)
+{
+	errno = 0;
+	const long temp = strtol(arg, NULL, 10);
+	if (errno || temp <= 0) {
+		fprintf(stderr, "error arg:%c %s\n", (char)key, arg);
+		argp_usage(state);
+	}
+
+	return temp;
+}
+
+error_t argp_parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args = 0;
+
+	switch (key) {
+	case 'p':
+		env.pid = atoi(arg);
+		break;
+	case 't':
+		env.trace_all = true;
+		break;
+	case 'a':
+		env.show_allocs = true;
+		break;
+	case 'o':
+		env.min_age_ns = 1e6 * atoi(arg);
+		break;
+	case 'c':
+		strncpy(env.command, arg, sizeof(env.command) - 1);
+		break;
+	case 'C':
+		env.combined_only = true;
+		break;
+	case 'F':
+		env.wa_missing_free = true;
+		break;
+	case 's':
+		env.sample_rate = argp_parse_long(key, arg, state);
+		break;
+	case 'T':
+		env.top_stacks = atoi(arg);
+		break;
+	case 'z':
+		env.min_size = argp_parse_long(key, arg, state);
+		break;
+	case 'Z':
+		env.max_size = argp_parse_long(key, arg, state);
+		break;
+	case 'O':
+		strncpy(env.object, arg, sizeof(env.object) - 1);
+		break;
+	case 'P':
+		env.percpu = true;
+		break;
+	case ARGP_KEY_ARG:
+		pos_args++;
+
+		if (pos_args == 1) {
+			env.interval = argp_parse_long(key, arg, state);
+		}
+		else if (pos_args == 2) {
+			env.nr_intervals = argp_parse_long(key, arg, state);
+		} else {
+			fprintf(stderr, "Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
```

Comment:
static, same for a bunch of other funcs

Reply:
earlier in the file, the prototypes are static